tensorboard/plugins/graph/graphs_demo.py

```diff
@@ -0,0 +1,133 @@
+
+    The `tagged` run has a top-level run graph as well as steps
+    `step_0000` through `step_0002`, each with profile data.
+    """
+    logdir = os.path.join(LOGDIR, "tagged")
+    with tf.compat.v1.Graph().as_default():
+        with tf.compat.v1.Session() as sess:
+            step_tensor = tf.compat.v1.placeholder(shape=(), dtype=tf.int32)
+            output = step_tensor * 2
+
+            writer = tf.compat.v1.summary.FileWriter(logdir)
+            with writer:
+                writer.add_graph(sess.graph)
+                for step in range(3):
+                    feed_dict = {step_tensor: step}
+                    run_options = tf.compat.v1.RunOptions()
+                    run_options.trace_level = tf.compat.v1.RunOptions.FULL_TRACE
+                    run_metadata = tf.compat.v1.RunMetadata()
+                    s = sess.run(
+                        output,
+                        feed_dict=feed_dict,
+                        options=run_options,
+                        run_metadata=run_metadata,
+                    )
+                    writer.add_run_metadata(run_metadata, "step_%04d" % step)
+
+
+def keras():
+    """Create a Keras conceptual graph and op graphs.
+
+    The `keras/train` run has a run-level graph, a `batch_2` tag with op
+    graph only (`graph_run_metadata_graph` plugin), and a `keras` tag
+    with a Keras conceptual graph only (`graph_keras_model` plugin).
+    """
+    logdir = os.path.join(LOGDIR, "keras")
+
+    data_size = 1000
+    train_fac = 0.8
+    train_size = int(data_size * train_fac)
+    x = np.linspace(-1, 1, data_size)
+    np.random.shuffle(x)
+    y = 0.5 * x + 2 + np.random.normal(0, 0.05, (data_size,))
+    (x_train, y_train) = x[:train_size], y[:train_size]
+    (x_test, y_test) = x[train_size:], y[train_size:]
+
+    layers = [
+        tf.keras.layers.Dense(16, input_dim=1),
+        tf.keras.layers.Dense(1),
+    ]
+    model = tf.keras.models.Sequential(layers)
+    model.compile(
+        loss=tf.keras.losses.mean_squared_error,
+        optimizer=tf.keras.optimizers.SGD(lr=0.2),
+    )
+    model.fit(
+        x_train,
+        y_train,
+        batch_size=train_size,
+        verbose=0,
+        epochs=100,
+        validation_data=(x_test, y_test),
+        callbacks=[tf.keras.callbacks.TensorBoard(logdir)],
+    )
+
+
+def profile():
+    """Create data with op graphs and profile data.
+
+    The `profile` run has tags `prof_f` with both profile and op graph data
+    (`graph_run_metadata` plugin), and `prof_g` with profile data only
+    (`graph_run_metadata_graph` plugin).
+    """
+
+    logdir = os.path.join(LOGDIR, "profile")
+
+    @tf.function
+    def f(i):
+        return tf.constant(i) + tf.constant(i)
+
+    @tf.function
+    def g(i):
+        return tf.constant(i) * tf.constant(i)
+
+    with tf.summary.create_file_writer(logdir).as_default():
+        for step in range(3):
+            tf.summary.trace_on(profiler=True)
+            print(f(step).numpy())
```

Comment:
I think for more correct looking graph, you need to do `tf.constant(step)` here and not pass Python number as input to tf.function.

Reply:
That doesn’t work:

```diff
diff --git a/tensorboard/plugins/graph/graphs_demo.py b/tensorboard/plugins/graph/graphs_demo.py
index 2329b2581..38dd41cb4 100644
--- a/tensorboard/plugins/graph/graphs_demo.py
+++ b/tensorboard/plugins/graph/graphs_demo.py
@@ -123,3 +123,3 @@ def profile():
             tf.summary.trace_on(profiler=True)
-            print(f(step).numpy())
+            print(f(tf.constant(step)).numpy())
             tf.summary.trace_export("prof_f", step=step, profiler_outdir=logdir)
```

```
TypeError: in user code:

    /HOMEDIR/.cache/bazel/_bazel_wchargin/52a95bbdd50941251730eb33b7476a66/execroot/org_tensorflow_tensorboard/bazel-out/k8-opt/bin/tensorboard/plugins/graph/graphs_demo.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/graph/graphs_demo.py:115 f  *
        return tf.constant(i) + tf.constant(i)
    /VIRTUAL_ENV/lib/python3.8/site-packages/tensorflow/python/framework/constant_op.py:264 constant  **
        return _constant_impl(value, dtype, shape, name, verify_shape=False,
    /VIRTUAL_ENV/lib/python3.8/site-packages/tensorflow/python/framework/constant_op.py:281 _constant_impl
        tensor_util.make_tensor_proto(
    /VIRTUAL_ENV/lib/python3.8/site-packages/tensorflow/python/framework/tensor_util.py:457 make_tensor_proto
        _AssertCompatible(values, dtype)
    /VIRTUAL_ENV/lib/python3.8/site-packages/tensorflow/python/framework/tensor_util.py:334 _AssertCompatible
        raise TypeError("Expected any non-tensor type, got a tensor instead.")
```

As written, the graph looks okay to me:

![Screenshot of a graph with `Const` and `Const_1` inputs to `add`][ss]

[ss]: https://user-images.githubusercontent.com/4317806/102144079-b6635600-3e19-11eb-8228-b6665b75998b.png

And I don’t think that it matters too much what exactly the graph looks
like; I’m mostly trying to check that the data gets plumbed through
properly. It would be cool to have a set of demos as test cases for
weird graph rendering issues, but this isn’t meant to fill that need.